### PR TITLE
feat: parse nested pydantic models `fetchone` and `fetchall` + add shortcuts for insert_query and update_query into `Database`

### PR DESCRIPTION
**Pull request by dni**
disaster pr :) turned into a temporary new dev branch

### CODE REVIEW

File: lnbits/app.py [hunk 1]
```diff
@@ -304,7 +308,6 @@ async def check_installed_extension_files(ext: InstallableExtension) -> bool:
 
 
 async def restore_installed_extension(app: FastAPI, ext: InstallableExtension):
-    await add_installed_extension(ext)
```
Comment (dni) [inline lnbits/app.py:307]: .

File: lnbits/app.py [hunk 2]
```diff
@@ -240,7 +238,8 @@ async def check_installed_extensions(app: FastAPI):
                 )
         except Exception as e:
             logger.warning(e)
-            await deactivate_extension(ext.id)
+            settings.deactivate_extension_paths(ext.id)
```
Comment (motorina0) [inline lnbits/app.py:241]: wrong

File: lnbits/core/crud.py [hunk 3]
```diff
@@ -131,52 +62,45 @@ async def delete_account(user_id: str, conn: Optional[Connection] = None) -> Non
...
 async def get_account(
     user_id: str, conn: Optional[Connection] = None
-) -> Optional[User]:
-    row = await (conn or db).fetchone(
-        """
-           SELECT id, email, username, pubkey, created_at, updated_at, extra
-           FROM accounts WHERE id = :id
-        """,
+) -> Optional[Account]:
+    return await (conn or db).fetchone(
```
Comment (motorina0) [inline lnbits/core/crud.py:98]: why not make a `db.select()` similar to `db.update()`?

File: lnbits/core/crud.py [hunk 4]
```diff
@@ -197,215 +121,89 @@ async def delete_accounts_no_wallets(
...
-            FROM wallets
-            WHERE "user" = :user and wallets.deleted = false
-            """,
-            {"user": user_id},
-        )
-    else:
-        return None
 
+async def get_user(
+    account_or_id: Union[Account, str], conn: Optional[Connection] = None
```
Comment (motorina0) [inline lnbits/core/crud.py:165]: looks hackish
better to have two different optional parmeters

File: lnbits/core/crud.py [hunk 5]
```diff
@@ -463,93 +250,75 @@ async def drop_extension_db(*, ext_id: str, conn: Optional[Connection] = None) -
...
+    user_extension: UserExtension, conn: Optional[Connection] = None
 ) -> None:
-    await (conn or db).execute(
-        """
-        INSERT INTO extensions ("user", extension, active) VALUES (:user, :ext, :active)
-        ON CONFLICT ("user", extension) DO UPDATE SET active = :active
-        """,
-        {"user": user_id, "ext": extension, "active": active},
-    )
+    where = """extension = :extension AND "user" = :user"""
```
Comment (motorina0) [inline lnbits/core/crud.py:307]: shouldn't there be a `WHERE`?

File: lnbits/core/crud.py [hunk 6]
```diff
@@ -1235,9 +1013,12 @@ async def create_admin_settings(super_user: str, new_settings: dict):
 
 # db versions
 # --------------
-async def get_dbversions(conn: Optional[Connection] = None):
-    rows = await (conn or db).fetchall("SELECT * FROM dbversions")
-    return {row["db"]: row["version"] for row in rows}
```
Comment (motorina0) [inline lnbits/core/crud.py:1240]: what was wrong with the old way?

File: lnbits/core/extensions/extension_manager.py [hunk 7]
```diff
@@ -3,37 +3,38 @@
...
     await ext_info.download_archive()
 
     ext_info.extract_archive()
 
-    db_version = (await get_dbversions()).get(ext_info.id, 0)
+    db_version = (await get_dbversions()).get(ext_id, 0)
     await migrate_extension_database(extension, db_version)
 
-    await add_installed_extension(ext_info)
+    await create_installed_extension(ext_info)
```
Comment (motorina0) [inline lnbits/core/extensions/extension_manager.py:33]: what if it already exists?

File: lnbits/core/migrations.py [hunk 8]
```diff
@@ -553,3 +551,36 @@ async def m022_add_pubkey_to_accounts(db):
         await db.execute("ALTER TABLE accounts ADD COLUMN pubkey TEXT")
     except OperationalError:
         pass
+
+
+async def m023_add_column_column_to_apipayments(db):
+    """
+    renames hash to payment_hash and drops unused index
+    """
+    await db.execute("DROP INDEX by_hash")
```
Comment (motorina0) [inline lnbits/core/migrations.py:560]: does not re-create the index?

File: lnbits/core/models.py [hunk 9]
```diff
@@ -238,29 +270,38 @@ class CreatePayment(BaseModel):
...
     preimage: Optional[str] = None
-    expiry: Optional[datetime.datetime] = None
+    expiry: Optional[datetime] = None
     extra: Optional[dict] = None
     webhook: Optional[str] = None
     fee: int = 0
 
 
-class Payment(FromRowModel):
+# TODO: thanks about paymeny extra
```
Comment (motorina0) [inline lnbits/core/models.py:279]: you're welcome

File: lnbits/core/crud.py [hunk 10]
```diff
@@ -1139,30 +918,29 @@ async def check_internal(
...
-        WHERE hash = :hash AND amount > 0
+        WHERE payment_hash = :payment_hash AND amount > 0
         """,
-        {"hash": payment_hash},
+        {"payment_hash": payment_hash},
     )
     if not row:
         return True
-    return row["status"] == PaymentState.PENDING.value
+    return row["status"] == PaymentState.SUCCESS.value
```
Comment (motorina0) [inline lnbits/core/crud.py:936]: function should be renamed to something that like `is_internal_status_success`

File: lnbits/core/services.py [hunk 11]
```diff
@@ -762,7 +766,7 @@ async def check_admin_settings():
             send_admin_user_to_saas()
 
         account = await get_account(settings.super_user)
-        if account and account.config and account.config.provider == "env":
+        if account and account.extra and account.extra.provider == "env":
```
Comment (motorina0) [inline lnbits/core/services.py:769]: nitpick:
 - add method to account
```py
account.has_provider("env")
```

File: lnbits/core/services.py [hunk 12]
```diff
@@ -809,57 +813,64 @@ def update_cached_settings(sets_dict: dict):
...
 async def init_admin_settings(super_user: Optional[str] = None) -> SuperSettings:
+    async def new_account(account_id: str) -> Account:
+        now = datetime.now()
+        account = Account(
+            id=account_id,
+            extra=UserExtra(provider="env"),
+            created_at=now,
+            updated_at=now,
+        )
+        await create_account(account)
```
Comment (motorina0) [inline lnbits/core/services.py:824]: should create default user also?
maybe use `create_user_account`

Comment (motorina0) [reply lnbits/core/services.py:824]: move line 832 here

File: lnbits/core/migrations.py (hunk 8 above)
Comment (dni) [reply lnbits/core/migrations.py:560]: in `m025`

Comment (dni) [reply lnbits/core/migrations.py:560]: done

File: lnbits/core/models.py [hunk 13]
```diff
@@ -27,6 +27,15 @@
 )
 
 
+def json_custom_serialization(_, o):
+    if isinstance(o, datetime):
+        return o.isoformat()
+    raise TypeError(f"Object is not JSON serializable: {o}")
+
+
+json.JSONEncoder.default = json_custom_serialization  # type: ignore[method-assign]
```
Comment (dni) [inline lnbits/core/models.py:36]: @motorina0 using `user.json()` inside the generic api makes this unecessary

Comment (dni) [reply lnbits/core/models.py]: removed again

File: lnbits/core/services.py (hunk 11 above)
Comment (dni) [reply lnbits/core/services.py:768]: maybe if its used more often

File: lnbits/core/services.py [hunk 14]
```diff
@@ -697,38 +687,33 @@ def fee_reserve_total(amount_msat: int, internal: bool = False) -> int:
...
-                "wallet_balance": wallet.balance,
-                "payment": payment.dict(),
-            }
-        ),
-    )
-
-    await websocket_updater(
-        payment.payment_hash, json.dumps({"pending": payment.pending})
+    await websocket_manager.send_data(payment.json(), wallet.inkey)
+    # json.dumps(
```
Comment (dni) [inline lnbits/core/services.py:691]: !

File: lnbits/static/css/base.css [hunk 15]
```diff
@@ -528,7 +528,9 @@ video {
 
 .q-card.q-card--dark,
 .q-date--dark {
-  box-shadow: none;
+  border: 1px solid rgba(0,0,0,.12);
```
Comment (dni) [inline lnbits/static/css/base.css]: @arcbtc i think you need to add this to `base.scss` the base.css will be overwritten by `make bundle`

File: lnbits/static/js/account.js [hunk 16]
```diff
@@ -63,6 +64,27 @@ window.app = Vue.createApp({
         this.$q.localStorage.set('lnbits.gradientBg', false)
       }
     },
+    applyBorder: function () {
```
Comment (motorina0) [inline lnbits/static/js/account.js:67]: uff, NOK
can we move the `CSS` to its own file please and use class names

File: lnbits/core/views/payment_api.py [hunk 17]
```diff
@@ -145,60 +143,42 @@ async def api_payments_create_invoice(data: CreateInvoice, wallet: Wallet):
...
+                        payment.extra["lnurl_response"] = True
             except (httpx.ConnectError, httpx.RequestError) as ex:
                 logger.error(ex)
-                lnurl_response = False
+                payment.extra["lnurl_response"] = False
 
-    return {
-        "payment_hash": invoice.payment_hash,
-        "payment_request": payment_request,
-        "lnurl_response": lnurl_response,
```
Comment (motorina0) [inline lnbits/core/views/payment_api.py:198]: - `lnurl_response` is not in extra which breaks backwards compability

File: lnbits/core/crud.py (hunk 5 above)
Comment (dni) [reply lnbits/core/crud.py]: done

File: lnbits/core/views/payment_api.py (hunk 17 above)
Comment (dni) [reply lnbits/core/views/payment_api.py:198]: add to breaking change readme

File: lnbits/core/views/payment_api.py [hunk 18]
```diff
@@ -330,19 +302,14 @@ async def api_payments_pay_lnurl(
...
+
+    payment = await pay_invoice(
         wallet_id=wallet.wallet.id,
         payment_request=params["pr"],
         description=data.description,
         extra=extra,
     )
-
-    return {
-        "success_action": params.get("successAction"),
```
[the remark this answers is not in the document]
Comment (dni) [reply lnbits/core/views/payment_api.py:341]: add to breaking change readme

File: lnbits/app.py (hunk 1 above)
Comment (dni) [reply lnbits/app.py:307]: its replaced by `    await update_installed_extension_state(ext_id=ext.id, active=True)
`

File: lnbits/app.py (hunk 2 above)
Comment (dni) [reply lnbits/app.py]: reverted, is think the `deactivate_extension` is unecessary its just those 2 lines and used twice, i thought its easier to read and the the actual db action behind it immediatly. but i dont mind...

File: lnbits/core/extensions/extension_manager.py (hunk 7 above)
Comment (dni) [reply lnbits/core/extensions/extension_manager.py]: fixed

File: lnbits/static/js/account.js [hunk 19]
```diff
@@ -63,6 +64,27 @@ window.app = Vue.createApp({
...
+      slef = this
```
Comment (dni) [inline lnbits/static/js/account.js:68]: self is not needed

File: lnbits/static/js/base.js [hunk 20]
```diff
@@ -517,6 +521,30 @@ window.windowMixin = {
         document.head.appendChild(style)
       }
     },
+    applyBorder: function () {
```
Comment (dni) [inline lnbits/static/js/base.js:524]: this code is copy pastes from  lnbits/static/js/account.js

File: lnbits/static/js/wallet.js [hunk 21]
```diff
@@ -142,9 +145,11 @@ window.app = Vue.createApp({
         )
         .then(response => {
           this.receive.status = 'success'
-          this.receive.paymentReq = response.data.payment_request
+          this.receive.paymentReq = response.data.bolt11
           this.receive.paymentHash = response.data.payment_hash
 
+          // TODO: lnurl_callback and lnurl_response
```
Comment (dni) [inline lnbits/static/js/wallet.js:151]: TODO

**Review by dni — COMMENTED** [review]
reviewed again, some stuff is still in TODO and a bit weird but i think rather merge it now than getting even bigger

**Review by motorina0 — APPROVED** [review]

**Pull request merged**